### PR TITLE
TypeScript: add script to run all TypeScript tests

### PR DESCRIPTION
**Pull request by azz**
Runs prettier on all TypeScript tests from `../TypeScript/tests`.

Groups error by type, prints counts, and writes the full details to files, one file per type of error.

run-external-typescript-tests

Output (too fast to see in the gif):

```
$ node scripts/run-external-typescript-tests.js
13328 good, 1757 bad
13328/15085 files printed without errors.

624     ast(input) !== ast(prettier(input))
312     Error: Value undefined is not a valid document
110     Error: Comment "<omitted>" was not printed. Please report this error!
90      SyntaxError: ';' expected.
88      SyntaxError: Expression expected.
83      TypeError: Cannot read property 'type' of undefined
55      prettier(input) !== prettier(prettier(input))
46      SyntaxError: Invalid character.
44      SyntaxError: ',' expected.
35      SyntaxError: Identifier expected.
24      SyntaxError: Declaration or statement expected.
21      SyntaxError: '(' expected.
19      SyntaxError: Type expected.
17      SyntaxError: ':' expected.
16      SyntaxError: Hexadecimal digit expected.
16      SyntaxError: Property or signature expected.
12      SyntaxError: '=' expected.
12      SyntaxError: Unterminated template literal.
10      SyntaxError: Parameter declaration expected.
8       SyntaxError: 'super' must be followed by an argument list or member access.
8       SyntaxError: An extended Unicode escape value must be between 0x0 and 0x10FFFF inclusive.
8       SyntaxError: ')' expected.
7       SyntaxError: Unexpected token. A constructor, method, accessor, or property was expected.
6       SyntaxError: '}' expected.
6       TypeError: Cannot read property 'some' of undefined
6       SyntaxError: '{' expected.
6       SyntaxError: Property assignment expected.
5       SyntaxError: '{' or ';' expected.
5       SyntaxError: Unterminated Unicode escape sequence.
4       SyntaxError: Unterminated string literal.
4       SyntaxError: Digit expected.
3       SyntaxError: '=>' expected.
3       SyntaxError: Argument expression expected.
3       SyntaxError: Unterminated regular expression literal.
3       SyntaxError: 'as' expected.
3       SyntaxError: ']' expected.
3       Error: Comment location overlaps with node location
3       SyntaxError: An unary expression with the '-' operator is not allowed in the left-hand side of an exponentiation expression. Consider enclosing the expression in parentheses.
3       SyntaxError: Merge conflict marker encountered.
2       SyntaxError: Array element destructuring pattern expected.
2       SyntaxError: '>' expected.
2       SyntaxError: Enum member expected.
2       SyntaxError: Declaration expected.
2       SyntaxError: Variable declaration expected.
2       SyntaxError: Expression or comma expected.
2       SyntaxError: '*/' expected.
1       SyntaxError: An unary expression with the 'typeof' operator is not allowed in the left-hand side of an exponentiation expression. Consider enclosing the expression in parentheses.
1       SyntaxError: Binary digit expected.
1       SyntaxError: An unary expression with the 'delete' operator is not allowed in the left-hand side of an exponentiation expression. Consider enclosing the expression in parentheses.
1       SyntaxError: JSX element 'number' has no corresponding closing tag.
1       SyntaxError: Invalid 'reference' directive syntax.
1       SyntaxError: Type parameter declaration expected.
1       SyntaxError: 'case' or 'default' expected.
1       SyntaxError: Cannot read property 'pos' of undefined (undefined:undefined)
1       SyntaxError: JSX element 'T' has no corresponding closing tag.
1       SyntaxError: Unexpected end of text.
1       SyntaxError: 'of' expected.
1       SyntaxError: An AMD module cannot have multiple name assignments.
Writing errors to './errors/' directory
```

### CODE REVIEW

File: bin/prettier.js [hunk 1]
```diff
@@ -153,20 +153,21 @@ function format(input) {
     const pp = prettier.format(input, options);
     const pppp = prettier.format(pp, options);
     if (pp !== pppp) {
-      process.stdout.write("\n");
-      console.error('prettier(input) !== prettier(prettier(input))');
-      console.error(diff(pp, pppp));
-      process.exitCode = 2;
+      throw "prettier(input) !== prettier(prettier(input))\n" + diff(pp, pppp);
```
Comment (azz) [inline bin/prettier.js]: * Throwing here so we log the filename in the error (`handleError`).
* Throwing a string so we don't get the stack trace.

File: bin/prettier.js [hunk 2]
```diff
@@ -153,20 +153,21 @@ function format(input) {
...
       const past = cleanAST(prettier.__debug.parse(pp, options));
 
       if (ast !== past) {
-        process.stdout.write("\n");
-        console.error('ast(input) !== ast(prettier(input))');
-        console.error(diff(ast, past));
-        console.error(diff(input, pp));
-        process.exitCode = 2;
+        const MAX_AST_SIZE = 2097152; // 2MB
+        const astDiff = (ast.length > MAX_AST_SIZE || past.length > MAX_AST_SIZE)
```
Comment (azz) [inline bin/prettier.js]: Ran in to an issue with this case:
 * `../TypeScript/tests/cases/compiler/resolvingClassDeclarationWhenInBaseTypeResolution.ts`

The AST was several megabytes, and the the process was hanging (killed after one minute) on `diff(ast, past)`.

Does skipping diffs for huge ASTs seems reasonable?

Comment (vjeux) [inline bin/prettier.js]: Totally reasonable

**Comment by vjeux** [comment]
Thanks! Going to be super useful

**Pull request merged**